### PR TITLE
gh-120145: Speed up symlink lookup for `posixpath.realpath()`

### PR DESCRIPTION
**Pull request by nineteendo**
### Benchmark

<details><summary>Show script</summary>

```sh
# speedup-symlink-lookup.sh
mkdir -p "tmp/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/\
a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/\
a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/\
a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/\
a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/\
a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/\
a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/\
a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a"
echo 2 dirs
main/python.exe                   -m timeit -s "import os; path = os.getcwd() + '/tmp' + '/a' * 2"   "os.path.realpath(path)"
speedup-symlink-lookup/python.exe -m timeit -s "import os; path = os.getcwd() + '/tmp' + '/a' * 2"   "os.path.realpath(path)"
echo 20 dirs
main/python.exe                   -m timeit -s "import os; path = os.getcwd() + '/tmp' + '/a' * 20"  "os.path.realpath(path)"
speedup-symlink-lookup/python.exe -m timeit -s "import os; path = os.getcwd() + '/tmp' + '/a' * 20"  "os.path.realpath(path)"
echo 200 dirs
main/python.exe                   -m timeit -s "import os; path = os.getcwd() + '/tmp' + '/a' * 200" "os.path.realpath(path)"
speedup-symlink-lookup/python.exe -m timeit -s "import os; path = os.getcwd() + '/tmp' + '/a' * 200" "os.path.realpath(path)"

echo ---

ln -nsfw . tmp/b
echo 2 links
main/python.exe                   -m timeit -s "import os; path = os.getcwd() + '/tmp' + '/b' * 2"      "os.path.realpath(path)"
speedup-symlink-lookup/python.exe -m timeit -s "import os; path = os.getcwd() + '/tmp' + '/b' * 2"      "os.path.realpath(path)"
echo 20 links
main/python.exe                   -m timeit -s "import os; path = os.getcwd() + '/tmp' + '/b' * 20"     "os.path.realpath(path)"
speedup-symlink-lookup/python.exe -m timeit -s "import os; path = os.getcwd() + '/tmp' + '/b' * 20"     "os.path.realpath(path)"
echo 200 links
main/python.exe                   -m timeit -s "import os; path = os.getcwd() + '/tmp' + '/b' * 200"    "os.path.realpath(path)"
speedup-symlink-lookup/python.exe -m timeit -s "import os; path = os.getcwd() + '/tmp' + '/b' * 200"    "os.path.realpath(path)"
echo 2,000 links
main/python.exe                   -m timeit -s "import os; path = os.getcwd() + '/tmp' + '/b' * 2_000"  "os.path.realpath(path)"
speedup-symlink-lookup/python.exe -m timeit -s "import os; path = os.getcwd() + '/tmp' + '/b' * 2_000"  "os.path.realpath(path)"
echo 20,000 links
main/python.exe                   -m timeit -s "import os; path = os.getcwd() + '/tmp' + '/b' * 20_000" "os.path.realpath(path)"
speedup-symlink-lookup/python.exe -m timeit -s "import os; path = os.getcwd() + '/tmp' + '/b' * 20_000" "os.path.realpath(path)"
```

</details>

<details><summary>2 dirs - no difference</summary>

```none
10000 loops, best of 5: 35.2 usec per loop
10000 loops, best of 5: 35.2 usec per loop
```

</details>

<details><summary>20 dirs - no difference</summary>

```none
2000 loops, best of 5: 129 usec per loop
2000 loops, best of 5: 129 usec per loop
```

</details>

<details><summary>200 dirs - no difference</summary>

```none
100 loops, best of 5: 2.4 msec per loop
100 loops, best of 5: 2.39 msec per loop
```

</details>

---

<details><summary>2 links - 1.18x faster</summary>

```none
5000 loops, best of 5: 41 usec per loop
10000 loops, best of 5: 34.7 usec per loop
```

</details>

<details><summary>20 links - 3.31x faster</summary>

```none
2000 loops, best of 5: 131 usec per loop
10000 loops, best of 5: 39.6 usec per loop
```

</details>

<details><summary>200 links - 12.03x faster</summary>

```none
200 loops, best of 5: 1.05 msec per loop
5000 loops, best of 5: 87.3 usec per loop
```

</details>

<details><summary>2,000 links - 18.33x faster</summary>

```none
50 loops, best of 5: 9.77 msec per loop
500 loops, best of 5: 533 usec per loop
```

</details>

<details><summary>20,000 links - 19.27x faster</summary>

```none
5 loops, best of 5: 95.4 msec per loop
50 loops, best of 5: 4.95 msec per loop
```

</details>

* Issue: gh-120145

### CODE REVIEW

File: Lib/posixpath.py [hunk 1]
```diff
@@ -443,6 +443,17 @@ def _realpath(filename, strict=False, sep=sep, curdir=curdir, pardir=pardir,
             newpath = path + name
         else:
             newpath = path + sep + name
+        if newpath in seen:
```
Comment (nineteendo) [inline Lib/posixpath.py]: Is this faster for `strict` mode?

```suggestion
        if maxlinks is None and newpath in seen:
```

**Comment by nineteendo** [comment]
cc @barneygale

**Comment by picnixz** [comment]
Just wondering, but how many symlinks do you need until you see a noticable effect? (in your case it's 200_000 but I honestly don't think you'll ever have that use case or maybe I'm biased?)

**Comment by nineteendo** [comment]
> how many symlinks do you need until you see a noticable effect?

2 is enough for an 11% speedup.

> I honestly don't think you'll ever have that use case

This is partially a security issue, given that paths can get really long using symlinks.

**Comment by picnixz** [comment]
> 2 is enough for an 11% speedup.

IMO, this is enough to motivate your feature (it's not uncommon to have a chain `symlink1->symlink2->file` I think).

**Comment by nineteendo** [comment]
> it's not uncommon to have a chain `symlink1->symlink2->file` I think

Note that this only works for the same symlink. e.g. `symlink->.` and `symlink1/symlink1`.
But as we're already caching them, speedup up retrieval only makes sense, right?

**Comment by barneygale** [comment]
How much slower is this for paths that _don't_ involve symlinks?

**Comment by nineteendo** [comment]
Only 3% for 20 directories. <2% for other cases.

**Comment by barneygale** [comment]
Then IMHO it's not worth it - symlinks are much rarer than non-symlinks, and so even a 2-3% slowdown has an outsized effect.

**Comment by nineteendo** [comment]
Let me benchmark again with `seen and newpath in seen`, that's a bit faster.

**Comment by nineteendo** [comment]
I can't measure a slowdown anymore. For 20 directories it's faster now, but that's within the margin of error.

**Comment by nineteendo** [comment]
Barney, is it fine now or should I close this? The overhead is 8.2 nanoseconds, on something measured in microseconds:

```zsh
wannes@Stefans-iMac ~ % python -m timeit -s "seen = {}; key = 'foo'"
50000000 loops, best of 5: 6.5 nsec per loop
wannes@Stefans-iMac ~ % python -m timeit -s "seen = {}; key = 'foo'" "seen and key in seen"
20000000 loops, best of 5: 14.7 nsec per loop
```

**Comment by nineteendo** [comment]
@barneygale, I moved the `is None` check inside `if not name or name == curdir`, that means the performance decrease is only 3 nanoseconds now:

```none
wannes@Stefans-iMac ~ % python -m timeit -s "seen = {}; key = 'foo'" "seen is None"        
20000000 loops, best of 5: 11.7 nsec per loop
wannes@Stefans-iMac ~ % python -m timeit -s "seen = {}; key = 'foo'" "seen and key in seen"
20000000 loops, best of 5: 14.7 nsec per loop
```

File: Lib/posixpath.py [hunk 2]
```diff
@@ -428,13 +428,14 @@ def _realpath(filename, strict=False, sep=sep, curdir=curdir, pardir=pardir,
...
         if not name or name == curdir:
-            # current dir
-            continue
+            if name is None:
+                # resolved symlink target
+                seen[rest.pop()] = path
+                continue
+            else:
+                # current dir
+                continue
```
Comment (barneygale) [inline Lib/posixpath.py:438]: This change looks good.

**Review by barneygale — COMMENTED** [review]
I'm still -1 on speeding up symlinks at the expense of non-symlinks

File: Lib/posixpath.py (hunk 2 above)
Comment (nineteendo) [reply Lib/posixpath.py:438]: OK, I'll create a PR and close this one.

**Pull request closed without merging**